### PR TITLE
Wait to compute the neighborhood until after finalizing the beam.

### PR DESCRIPTION
**Pull request by mlmenard**

### CODE REVIEW

File: src/dachshund/beam.rs [hunk 1]
```diff
@@ -189,6 +191,7 @@ impl<'a, TGraph: GraphBase> Beam<'a, TGraph> {
                     scored_expansion_candidates.insert(ell);
                 }
             }
+            previous_candidates.insert(candidate.checksum.unwrap(), candidate);
```
Comment (bogdanstate) [inline src/dachshund/beam.rs:194]: a comment as to what this `previous_candidates` map does (i.e., being passed as hints?) would be quite useful.

**Review by bogdanstate — APPROVED** [review]
Looks good, although having a unit test to ensure `set_neighborhood_with_hint` works as expected would be quite useful!

File: src/dachshund/candidate.rs [hunk 2]
```diff
@@ -76,7 +80,8 @@ where
     max_core_node_edges: usize,
     ties_between_nodes: usize,
     local_guarantee: LocalDensityGuarantee,
-    neighborhood: HashMap<NodeId, usize>,
+    neighborhood: Option<HashMap<NodeId, usize>>,
+    recipe: Option<(Option<u64>, NodeId)>,
```
Comment (zellski) [inline src/dachshund/candidate.rs:84]: I've never been crazy about using unnamed tuples for anything, much less function parameters. It's SO easy to create at minimum a new type, if not an actual struct with field names (which is what I prefer).

However, the code base already has a ton of such tuples in it, I foresee either me just getting over this, or else doing a single replacement sweep of the whole project sometime.

Comment (mlmenard) [reply src/dachshund/candidate.rs:84]: I agree. I don't want more of these to accumulate. I'll switch this to a struct.

**Pull request merged**